wurst/event/DamageQueue.wurst

```diff
@@ -0,0 +1,108 @@
+package DamageQueue
+import DamageModify
+import LinkedList
+import TimerUtils
+
+/*
+	DamageQueue is created to make damage event listeners follow the same order
+	in which damage are invoked and also protect from the infinite thorns problem.
+
+	When damage is invoked, event listeners are run before target's health is
+	changed, which can lead to some hard to track bugs if there is nested damage
+	invokation, for example - DOTA's blademail.
+
+	Explaining example:
+
+	Test case:
+	init
+		addDamageListener() d ->
+			Log.info("Start " + (d castTo int).toString())
+		addDamageListener() d ->
+			if d.data > 0
+				d.getSource().damageTarget(d.getTarget(), d.getAmount() * d.data / (d.data+1), ATTACK_TYPE_NORMAL, d.data - 1).releaseInstance()
+			// Make damage resonance
+		addDamageListener() d ->
+			Log.info("End " + (d castTo int).toString())
+
+		// To demonstrate why is this even important
+		addDamageModifier() d ->
+			if d.getAmount() < d.getTarget().getMaxHP() / 2
+				d.preventDeath()
+				// Some tough guy's ability, who can't die from weak attacks
+
+	Then somewhere happens this:
+	unit2 has 101 hp out of 101.
+	unit1.damageTarget(unit2, 75, 2)
+	(Causing then 50 and 25 damage)
+	
+	Logs:
+	Start 1
+	Start 2
+	Start 3
+	End 3
+	*unit 2 is hit with 25 damage, 76 left*
+	End 2
+	*unit 2 is hit with 50 damage, 26 left*
+	End 1
+	*unit 2 is hit with 75 (26 actual) damage, unit 2 dies.*
+
+	If you replace .damageTarget with .damageTargetQueued, you'll get this
+	result with in same test case:
+
+	Start 1
+	End 1
+	*unit 2 is hit with 75 damage, 26 left*
+	Start 2
+	End 2
+	*unit 2 is hit with 50 (25.594 actual) damage, 0.406 left*
+	Start 3
+	End 3
+	*unit 2 is hit with 25 (0 actual) damage, unit 2 lives*
+
+	So the results in this test are different if we take order into consideration.
+
+	This is proved for war3 version 1.30.
+	As you can see, in second case damage processing follows the order given
+*/
```

Comment:
Why not just make the behavior of the queued damage the default? The second case seems like the correct way to implement this entire system. The first one seems confusing to me.

Reply:
The first one is the default of wc3. And since this snippet is not something simple that were in the game before *and* since it can be implemented other way, it is separated from the main system.
At least that's position I keep because of how @Frotty sees stdlib - every part of it should be as modular and simple as possible, and this conceptually different thing would make DMod too much complicated in one time.

After all, not everybody plays with thorns, so why would those bother themselves with needless overhead?

Reply:
The first one isn't default to WC3, it's default to how you structured your code.

I would argue that rather than every part of the std library being as simple as possible, that we should also prioritize the simplicity to use and the intuitiveness of the usage, even if that means some additional complexity in the implementation.

I think there's a way we can deal with nested damage calls elegantly, without increasing the overhead for the case where there are no nested calls.